PX4 protected build pub

jlaitine opened the pull request:
This PR implements PX4 running on NuttX protected build. The split between kernel and user space libraries is in top level:

kernel:
- All drivers
- Full uORB
- Parameters (but only loading at boot time)
- high resolution timer
- Anything that links directly to OS drivers or accesses hardware

user space:
- Another uORB, mostly only uORBManager communicating with kernel side over ioctl
- parameters
- hrt interafce, using an own highest priority task for callbacks

build scripts:
- Support for building nuttx kernel and user -side libraries, linking for this
- Support for generating "builtins" for kernel side modules
- Support for putting any module to eiither kernel or user side (adding "KERNEL" flag to CMakeLists.txt for the module). Restrictions apply; a module linking to nuttx kernel or accessing HW cannot be linked into user side..

LorenzMeier commented:
Thanks! I will set some quality time apart to go through it. This capability is one of the key reasons why we chose NuttX 10 years ago and why some of the PX4 fundamentals have been architected around file descriptors (to be able to go through kernel space cleanly). We have since optimized on various fronts, so I'm sure it needs some refresh and polishing.

dagar commented:
I'm very interested to dig into this and see exactly where (and how) you've split things between kernel and user space. I actually gave this a try with NuttX a few years ago, but quickly came to the conclusion the value was minimal if nearly all of PX4 was simply in user space (able to clobber itself).

jlaitine commented:
If someone has had time to look into this bunch of stuff; I'd appreciate comments on it; basically which parts are implemented in a proper manner, and which could be done better or doesn't make sense at all. I could start chopping it in pieces and make individual smaller PR:s, for example

dagar commented:
> If someone has had time to look into this bunch of stuff; I'd appreciate comments on it; basically which parts are implemented in a proper manner, and which could be done better or doesn't make sense at all. I could start chopping it in pieces and make individual smaller PR:s, for example


I'm still very interested in this and would like to start getting pieces in incrementally as soon as we get v1.12 out.

jlaitine commented:
Rebased against current master, + added the following:
on px4_fmuv5_protected target:
- removed the micrortps agent from the build, it is not relevant for this work
- changed flash and sram allocation between kernel and user space, doubled the kernel memories so that it is easier to drop more stuff to kernel
- moved "sensors" to kernel side, it saves a lot of CPU when all the sensor data doesn't need to go many times between kenel-userspace barrier
- minor cleanups

Didn't have time to clean up all the things I had in mind yet though

jlaitine commented:
Just a re-base to keep this alive. The following items should be resolved (at least):
- couple of "command not found" nsh errors on other targets than "protected-build"; "userspace_init" systemcmd called from rcS, which only exists for protected, and "kparam" for loading kernel side parameters at boot
- all uORBs are allocated from user-side heap for protected-build, because of uORB performance optimization by inlining functions; these should be re-designed, especially the "advertised()" call. But didn't want to change the functionality of uORB for now, it is complicated enough
- one nuttx backport needed (protected build has broken in nuttx meanwhile)
- need a nicer way to select what goes to kernel side and what to user side. Now the selection is done by adding keyword "KERNEL" into component's CMakeLists.txt. But it would be better to have a separate kernel side component list with the "board" configuration

dagar commented:
Thanks @jlaitine, I'd love to start getting this in piece by piece. Could we find some time to discuss more specifically how to move forward incrementally?

I haven't reviewed everything in detail yet, but so far I generally agree with most of the decisions you've made, except for a few things like the user space HRT. What do you think about more or less keeping it entirely kernel side and limit users to ScheduledWorkItem, etc?

> all uORBs are allocated from user-side heap for protected-build, because of uORB performance optimization by inlining functions; these should be re-designed, especially the "advertised()" call. But I didn't want to change the functionality of uORB for now, it is complicated enough

What do you ultimately want to do here?

> one nuttx backport needed (protected build has broken in nuttx meanwhile)

Can we get this upstreamed and backported to the PX4 NuttX fork?

> need a nicer way to select what goes to kernel side and what to user side. Now the selection is done by adding keyword "KERNEL" into component's CMakeLists.txt. But it would be better to have a separate kernel side component list with the "board" configuration

We're very close to switching the cmake board definition over to kconfig. https://github.com/PX4/PX4-Autopilot/pull/17100

jlaitine commented:
> Thanks @jlaitine, I'd love to start getting this in piece by piece. Could we find some time to discuss more specifically how to move forward incrementally?

Yes, let's do. I am not sure if incrementally getting it in is a good way, since the pieces are not functional stand-alone, so it is bound to break before all pieces get in :)

> 
> I haven't reviewed everything in detail yet, but so far I generally agree with most of the decisions you've made, except for a few things like the user space HRT. What do you think about more or less keeping it entirely kernel side and limit users to ScheduledWorkItem, etc?

I'd keep it for now, to keep things simple. The hrt interface is used everywhere, just grep e.g. from src/modules. It is anyhow just a call-gate to kernel side, and in case of callbacks kernel just adding items into user-side work queue. A good step forward would be having some real OS/standard interface for the same functionality, but there is nothing quite similar there.

> 
> > all uORBs are allocated from user-side heap for protected-build, because of uORB performance optimization by inlining functions; these should be re-designed, especially the "advertised()" call. But I didn't want to change the functionality of uORB for now, it is complicated enough
> 
> What do you ultimately want to do here?

Currently publishers and subscribers fetch some data, such as "advertised" flag directly from uORBDeviceNode:s, which I think should live entirely in kernel side memory. These checks can't be call-gated nicely, because they are used so frequently. Optimization could maybe done in a way that these properties live directly on subscribers/publishers and is pushed from kernel to these when the status changes... or maybe there is some better solution.

Anyhow, this kind of details could be solved separately when there is compiling&working system in place.
> 
> > one nuttx backport needed (protected build has broken in nuttx meanwhile)
> 
> Can we get this upstreamed and backported to the PX4 NuttX fork?

It is in NuttX upstream already:
commit 1b75b5d5aa8c4924bc118cd3fd294984094857b8    Fix compilation of arm protected build

Would just need a cherry pick to PX4 branch

> 
> > need a nicer way to select what goes to kernel side and what to user side. Now the selection is done by adding keyword "KERNEL" into component's CMakeLists.txt. But it would be better to have a separate kernel side component list with the "board" configuration
> 
> We're very close to switching the cmake board definition over to kconfig. #17100

Right, this part probably needs to be done then. I will keep the current way for now

jlaitine commented:
btw; the build system gives me tons of headache (I've spent more hours with cmake & linking than coding this PR). I didn't get "make tests" to link properly, because px4_platfrom links directly to drivers_board. I just made another hack to get around that in topmost commit https://github.com/PX4/PX4-Autopilot/pull/16931/commits/1d4facec15ed62220080d31ad34d1f4269663c2c

If someone understands this better, please help with the cmake ;)

dagar commented:
> If someone understands this better, please help with the cmake ;)

I was hoping upstream NuttX would fully move to cmake at some point and we'd have an opportunity to simplify quite a bit, but that effort seems to have stalled.

For now let me know if you have any specific questions and I'll see what I can help with.

dagar commented:
> I'd keep it for now, to keep things simple. The hrt interface is used everywhere, just grep e.g. from src/modules. It is anyhow just a call-gate to kernel side, and in case of callbacks it is propagated through a user-space thread. A good step forward would be having some real OS/standard interface for the same functionality, but there is nothing quite similar there.

In userspace all that's really needed directly is `hrt_absolute_time()`, which we could replace with an appropriate call (or a wrapper short term).

dagar commented:
The test rack fails if anything reports error during initial bootup. http://px4-jenkins.dagar.ca:8080/blue/organizations/jenkins/PX4-Autopilot/detail/PR-16931/7/pipeline/143

![Screenshot from 2021-09-07 15-06-42](https://user-images.githubusercontent.com/84712/132397870-4cf24218-4b4a-4861-bb92-933635bf5957.png)

So this needs to be skipped in the regular builds.

jlaitine commented:
> The test rack fails if anything reports error during initial bootup. http://px4-jenkins.dagar.ca:8080/blue/organizations/jenkins/PX4-Autopilot/detail/PR-16931/7/pipeline/143

> So `kparam` needs to be skipped in the regular builds.

Thanks!

What about "command not found" message? There is one also from "userspace_init". 

possible ways to fix the kparam error:
- just remove the error (but this would leave the "command not found" message)
- stub the kparam for normal builds
- make "param load" just load the parameters on both kernel and user sides for protected build. Then this change is not needed in rcS

I'll try to fix these issues in some clean way if I can find some time for it today.

jlaitine commented:
fixed the issues with startup scripts throwing errors for other targets. The startup scripts are now unmodifed

- made a small addition to systemcmd/param to call the kernel side counterpart when loading the parameters -> no need to call it explicitly in startup scripts any more
- wrapped px4 userspace initialization and nsh_start into a new entrypoint function. The px4 userspace is initialized first, and only after that nsh is launched. Then there is no need for separate systemcmd for initializing the px4 userspace (which was silly quick-and-dirty hack in the first place)

jlaitine commented:
> > I'd keep it for now, to keep things simple. The hrt interface is used everywhere, just grep e.g. from src/modules. It is anyhow just a call-gate to kernel side, and in case of callbacks it is propagated through a user-space thread. A good step forward would be having some real OS/standard interface for the same functionality, but there is nothing quite similar there.
> 
> In userspace all that's really needed directly is `hrt_absolute_time()`, which we could replace with an appropriate call (or a wrapper short term).

Yes, I see what you mean. Currently the hrt_call_every is being used in logger (which I want to always keep in user side).

Also, the implementation of px4::ScheduledWorkItem is using the hrt interface directly; I didn't change the ScheduledWorkItem (it is a platform common function), but only linked it against the "userspace hrt interface" in userland.

So yes, I understand what you are after; but I don't see why we would limit the hrt interface usage for the "protected" build for some components, while the same components can use it in posix and flat builds?

There should be some other than memory protection reason to limit the usage of hrt, which is common to all?

jlaitine commented on src/drivers/px4io/CMakeLists.txt:
```diff
@@ -42,9 +42,11 @@ px4_add_module(
 	DEPENDS
 		arch_px4io_serial
 		circuit_breaker
-		mixer
```
This change is unintentional, will revert it

jlaitine commented on boards/px4/fmu-v6u/src/CMakeLists.txt:
```diff
@@ -40,6 +40,7 @@ if("${PX4_BOARD_LABEL}" STREQUAL  "bootloader")
 			nuttx_arch # sdio
 			nuttx_drivers # sdio
 			bootloader
+                        px4_layer # px4_platform_console_init
```
All of these, linking bootloaders to px4_platform is not valid any more; this has changed - will remove these

jlaitine commented on boards/px4/fmu-v5x/src/CMakeLists.txt:
```diff
@@ -43,7 +43,7 @@ add_library(drivers_board
 	usb.c
 )
-add_dependencies(drivers_board arch_board_hw_info platform_gpio_mcp23009)
+#add_dependencies(drivers_board arch_board_hw_info platform_gpio_mcp23009)
```
Don't understand/remember this one any more, will revisit

jlaitine commented:
Cleaned up some old/now unnecessary CMake changes

jlaitine commented on .gitmodules:
```diff
@@ -28,8 +28,8 @@
-	url = https://github.com/PX4/NuttX.git
-	branch = px4_firmware_nuttx-10.1.0+
+	url = https://github.com/tiiuae/nuttx.git
+	branch = protected_build
```
This will be removed as soon as https://github.com/PX4/PX4-Autopilot/pull/18212 is merged

davids5 replied:
#18212 is merged

jlaitine commented on boards/px4/fmu-v5/src/stm32_userspace.c:
```diff
@@ -0,0 +1,147 @@
+/****************************************************************************
```
All of this  could be moved to be common for all STM32 targets; however, would like to keep it here until there is another target that uses it

jlaitine commented on platforms/common/CMakeLists.txt:
```diff
@@ -52,11 +52,14 @@ add_library(px4_platform
+target_link_libraries(px4_platform prebuild_targets px4_work_queue)
+
+if("${PX4_BOARD}" MATCHES "sitl")
+	target_link_libraries(px4_platform drivers_board)
```
this is ugly, but without this the "make tests" won't link properly. I wonder if there is a cleaner solution?

jlaitine commented:
Dropped my own nuttx branch, it is no longer needed. This works on top of px4 nuttx directly

jlaitine commented:
Cleaned up the selection how to pick what goes to kernel and what goes to user side. This is now entirely done from within board and doesn't clobber every CMakeLists.txt file for modules/drivers etc.

jlaitine commented:
- re-based against master
- moved ekf2, mc_att_control and mc_rate_control to the kernel side for the example px4_fmuv5_protected target ( mostly to save CPU)
- desk-tested on holybro pixhawk4 hw

dagar commented:
I'm still seeing build failures.
![Screenshot from 2021-09-22 12-30-45](https://user-images.githubusercontent.com/84712/134384359-e5cd04ad-5a9e-4b40-85a2-bfa0c948a80c.png)

jlaitine commented:
> I'm still seeing build failures.
> ![Screenshot from 2021-09-22 12-30-45](https://user-images.githubusercontent.com/84712/134384359-e5cd04ad-5a9e-4b40-85a2-bfa0c948a80c.png)

The screenshot show error in cloning nuttx submodule. This PR doesn't touch any of the submodules. I just tried cloning to a clean directory, and building "make nxp_fmuk66-v3_test", getting no errors in cloning or building.

So I don't know what is wrong there, sorry.

jlaitine commented:
- Small fixes for whitespaces (tabs vs spaces) in CMake files

dagar commented on platforms/nuttx/NuttX/CMakeLists.txt:
```diff
@@ -269,6 +269,72 @@ endif()
 
+if (NOT CONFIG_BUILD_FLAT)
+
+set(nuttx_kernel_builtin_list)
```
indent

jlaitine replied:
right, fixed, thanks

dagar commented on src/modules/events/status_display.h:
```diff
@@ -45,7 +45,7 @@
 #include <uORB/Subscription.hpp>
 #include <uORB/topics/battery_status.h>
-#include <uORB/topics/cpuload.h>
+#include <uORB/topics/vehicle_cpuload.h>
```
If cpuload changes we'll need to update flight review. https://github.com/px4/flight_review

jlaitine replied:
Yes, this would be bad. I don't frankly remember what was the reason behind this any more; it was done very early when I checked if there was something HW dependent that could be simply replaced with os-provided interfaces. I'll check again if I can just remove this change.

jlaitine replied:
This is indeed not needed. Thanks for drawing my attention to it, I removed this change

dagar commented on src/modules/mavlink/CMakeLists.txt:
```diff
@@ -80,3 +76,18 @@ px4_add_module(
 	add_subdirectory(mavlink_tests)
 endif()
+
+if (NOT ${PX4_PLATFORM} STREQUAL "nuttx" OR CONFIG_BUILD_FLAT)
```
These could be replaced with directly publishing sensor_accel, sensor_gyro, etc.

jlaitine replied:
Yes, good point. I even have some implementation ready for this, but left it out in fear of breaking something. Was going to make another PR of it

jlaitine replied:
Do you want the generic, pure uORB based, implementation into this PR? Or should I make another PR before this one? Or would it be ok to continue with another PR later?

jlaitine replied:
@dagar I only now realized that you have already cleaned up the px4 driver libs to be pure uORB publishers. I just removed this change, and removed the linking to drivers__device from src/lib/drivers/accelerometer and src/lib/drivers/gyroscope CMakeLists.txt.

dagar commented on src/systemcmds/kparam/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+#
+############################################################################
+
+px4_add_module(
```
You could move this entirely to `systemcmds/param/CMakeLists.txt` and check for the protected build.

jlaitine replied:
Unfortunately, that doesn't work. The components are selected in board cmake files (protected.cmake + px4_add_board.cmake) by source directory, not by component name. This is the reason why this component is in another source directory, to be able to put them on other sides.

dagar commented on src/systemcmds/param/param.cpp:
```diff
@@ -216,12 +221,21 @@ param_main(int argc, char *argv[])
+
+#if defined (__PX4_NUTTX) && !defined (CONFIG_BUILD_FLAT) && !defined(__KERNEL__)
+			if (ret == 0) {
+				argv[0] = (char *)kparam_main_fn;
```
This is a bit awkward. Let me review the param and kernel param situation and see what else we can do.

We don't want to have to spend the next few years gradually cleaning up these CONFIG_BUILD_FLAT ifdefs.

jlaitine replied:
Agreed.

Again part of the thing that I wanted to do a top-down approach. First set up the framework and a working setup for a protected target, and later dig in to individual modules to clean up the interfaces. CONFIG_BUILD_FLAT ifdefs are found in those components which are on the userspace/kernelspace interface border.

There is always an option of putting these things in another source file, and adding a function call here, which is then stubbed away when doing normal "flat" build. I will just look into this again

jlaitine replied:
I moved all of this first into just one place in this file. Further it can be cleaned up by leving just a declaration into this file, and put the code in another source file? But having it in this same file is least amount of changes.

jlaitine commented:
Thanks @dagar for good comments! I will just keep on cleaning up this stuff

jlaitine commented:
Tried to clean up the points from @dagar. Tested flying just abit https://review.px4.io/plot_app?log=82780e52-a722-445c-8a9e-cf2e69c67b11 . The flight was indoors, the GPS position is garbage.

jlaitine commented:
re-base against master. Re-write of platforms/nuttx/src/px4/common/CMakeLists.txt

jlaitine commented:
small cleanups (indents etc.)

jlaitine commented:
Moved to draft while waiting for PR#17100 to land; there will be breakage from that. I will re-work some of the build files and mark this back to ready-for-review afterwards. Comments are of course still welcome!

jlaitine commented:
So this need of maintainer approval causes 24 hour delay for me to see if anything was wrong, as all the maintainers seem to be in the US (the approval comes earliest at the time when I call it a day). Is there something that could be done to this issue in CI?

jlaitine commented:
Nothing new; just a re-base to keep this alive.

The only remaining thing on my list is to do something reasonable with "parameters". It is still hackish. 

Parameters is just made to build and work with minimal changes so that nothing breaks for FLAT or posix builds.

It works also for protected build like this, but only on sdcard/filesystem (user space can't access flashparams or mtdparams directly) and if the parameter is used in kernel side module, there is always a reboot needed after a parameter change.

jlaitine commented:
And of course, since my every PR is my first one since months already, I have no idea whether this builds any more :)

jlaitine commented:
Converted back to draft. Picking things out of this as separate PRs

jlaitine commented:
re-base against current master

jlaitine commented on platforms/common/uORB/uORBManager.cpp:
```diff
@@ -103,6 +110,112 @@ uORB::DeviceMaster *uORB::Manager::get_device_master()
 	return _device_master;
 }
 
+#if defined (__PX4_NUTTX) && !defined (CONFIG_BUILD_FLAT) && defined(__KERNEL__)
```
This could be moved to a separate cpp file, and select in CMakeListst.txt ?

jlaitine commented:
Re-based this to keep it alive. Would still need to split this into smaller PRs; on the other hand, it is not possble to test any smaller parts alone

jlaitine commented:
I will try to further reduce size of this PR by picking some build and link changes out of it. Basically, adding support for compiling nuttx libraries for protected/kernel build, and changing link order: removing some circular dependencies and linking some of the libraries in later phase.

jlaitine commented:
Dropped everything that is not absolutely needed to compile and run this; also tested flying it.

jlaitine commented:
Just re-based to see if the weird CI fail goes away

jlaitine commented:
@dagar , @bkueng , this PR now has left the following
- initializing the userspace in boot
- some random boardctl things like reading vbus status into userspace
- boardctl call to launch modules in kernel side
- px4_fmu-v5_protected target, which is tested to build, flash and fly

Would you like me to further split this into smaller pieces, or is this reviewable as is? This brings in a working target.

bkueng commented on platforms/nuttx/src/px4/common/px4_userspace_init.cpp:
```diff
@@ -0,0 +1,20 @@
+#include <drivers/drv_hrt.h>
```
Missing copyright header

jlaitine replied:
Thanks, added

bkueng reviewed (approved):
You've made it through... with a lot of patience :D 
Nice work.

jlaitine commented:
Thank you for all the reviews and help with this! This is now a good base for gradual improvements in small steps, for which we already have some ideas.
